### PR TITLE
[4.0] Safer extensions deletion and better post install messages insertion in schema updates

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for no issue.

### Summary of Changes

With PR #22374 the usage of hard coded extension IDs has been removed. Thanks to @laoneo for this great work.

With that PR, statements for deleting old extensions in the schema update SQL scripts have been changed so they use the value of the element column in the extensions table  to identify the record to be deleted.

Unfortunately the element value is not unique, e.g. languages have same element name for admin and site, and so there is no unique constraint which keeps any 3rd party extensions from using same element name as other extensions do. And so using the element as the only criteria in the delete statements is theoretically unsafe. Practically all works as it is now with PR #22374 , but for future schema updates using the existing ones as a template for deleting extensions it might not be granted that they are safe.

This PR here extends the where clauses in the delete statements which have been modified by PR #22374 in schema updates by additional criteria which make them as safe as possible for the particular extension type.

In addition, it changes the insert statements in schema updates for new postinstall messages to the same statements as used for new installation, so the postinstall messages have the correct extension_id = the one of `files_joomla`.

### Testing Instructions

Code review.

@wilsonge and if possible @laoneo please review.

### Documentation Changes Required

None.

### CODE REVIEW

**Comment by richard67** [comment]
@franz-wohlkoenig Thanks for title update. Sorry I had forgotten the "[4.0] ".

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql [hunk 1]
```diff
@@ -1,6 +1,5 @@
...
-VALUES
-('PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_TITLE', 'PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_BODY', 'PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_ACTION', 'plg_system_httpheaders', 1, 'action', 'site://plugins/system/httpheaders/postinstall/introduction.php', 'httpheaders_postinstall_action', 'site://plugins/system/httpheaders/postinstall/introduction.php', 'httpheaders_postinstall_condition', '4.0.0', 1);
+INSERT INTO "#__postinstall_messages" ("extension_id", "title_key", "description_key", "action_key", "language_extension", "language_client_id", "type", "action_file", "action", "condition_file", "condition_method", "version_introduced", "enabled")
+SELECT "extension_id", 'PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_TITLE', 'PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_BODY', 'PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_ACTION', 'plg_system_httpheaders', 1, 'action', 'site://plugins/system/httpheaders/postinstall/introduction.php', 'httpheaders_postinstall_action', 'site://plugins/system/httpheaders/postinstall/introduction.php', 'httpheaders_postinstall_condition', '4.0.0', 1 FROM "#__extensions" WHERE "name" = 'files_joomla';
```
Comment (zero-24) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql:5]: this does not look correct to me. As this should be an insert statement

**Review by zero-24 — DISMISSED** [review]

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql (hunk 1 above)
Comment (richard67) [reply administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql:5]: It is the same statement as used in joomla.sql for new installation. It is an insert statement with a select for the values, "INSERT INTO ... SELECT ... FROM ...;". Quite a usual thing and used in joomla.sql many times. Works also well here in the schema update.

Comment (richard67) [reply administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql:5]: See e.g. here: [https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/sql/mysql/joomla.sql#L712](https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/sql/mysql/joomla.sql#L712)

Comment (richard67) [reply administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql:5]: Or here: [https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/sql/mysql/joomla.sql#L1589](https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/sql/mysql/joomla.sql#L1589)

Comment (richard67) [reply administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql:5]: @zero-24 You are wrong, sorry. See links posted above.

Comment (zero-24) [reply administrator/components/com_admin/sql/updates/postgresql/4.0.0-2017-10-10.sql:5]: hmm it still looks weird. But as i can't tests i want you to double check.

**Comment by richard67** [comment]
@zero-24 Thanks.

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**